# LCWA Powerpoint Dataset Exploration

This notebook uses python to explore the dataset of 1,000 ppt/x files
that were randomly identified and isolated from files archived by the
Library of Congress from US Federal government websites. 

In [23]:
import csv
#import json
import collections
import pandas as pd

_NB: this file is delimited by tabs and in UTF-16 text encoding. 
You will need to specify these things, as demonstrated below, in order
for the file's data to be processable by python._ 

In [15]:
linecount = 0
with open('lcwa_gov_powerpoint_metadata.csv', 'r', newline='', encoding='utf-16') as data:
    lcwa_powerpoint_metadata = csv.reader(data, dialect='excel-tab', delimiter='\t')
    for line in lcwa_powerpoint_metadata:
        print(linecount,' '.join(line))
        linecount += 1

0 urlkey timestamp original mimetype digest title company creation_date last_modified revision_number slide_count word_count file_size sha256 sha512
1 ada.ky.gov/adaemploymenttaxincentives.ppt 20100331231723 http://ada.ky.gov/ADAEMPLOYMENTTaxIncentives.ppt application/vnd.ms-powerpoint 4Y6Z6KRRHS3TIVQ2ZJVYVW7RSGXOBS6Y    Human Development Institute Promoting Independence, Productivity, and Integration for All People _________________________________    - 1998-12-17T14:31:23Z 2005-03-25T15:56:39Z 47 31 781 137216 9b6b74447622ac1646a538e63e452569e111cf775d1244373678d947405e33be 9567992755c0b1b38fdccb4440a4bb4b1758301b5cd1978b95bfe807babb70bf4c1e5ede85d9d26daee75196a3d62453ac9919bdbbf829508a26e72c35fc4df6
2 astrophysics.gsfc.nasa.gov/conferences/supernova1987a/staveley-smith.ppt 20090419154246 http://astrophysics.gsfc.nasa.gov/conferences/supernova1987a/Staveley-Smith.ppt application/vnd.ms-powerpoint 7EZSHVTKZE2FXP5SXI6DO2RJPIDLMXOP xNTD and the Pathway to SKA Lister Staveley-Smith - 200

695 cdph.ca.gov/programs/cpns/documents/network-spu-howfundingwork.ppt 20100316050604 http://www.cdph.ca.gov/programs/cpns/Documents/Network-SPU-HowFundingWork.ppt application/vnd.ms-powerpoint BT2LYDUELPJB6BSQ5VVRQHQSHK77R5TZ Where do Network Funds Come From and How are They Used?  2007-10-09T20:53:04Z 2007-10-10T14:42:32Z 1 1 85 847872 7bac474e6ddd1dcc3ead1ff60dcf1a47ef3560e537230992aa40936de6294299 68a9188d7e3f44ec7ef0cea167f529e492aa18ab6dbf002c8910ae851a05a8a88824fcd54fe2d9a85d65396b4e93bb98ba2b1520b4fc1dbc487e35569eb6abbc
696 gsics.nesdis.noaa.gov/pub/development/20150316/5k_aisehng_wu_iasi_modis_airs.ppt 20151118141038 https://gsics.nesdis.noaa.gov/pub/Development/20150316/5k_Aisehng_Wu_IASI_MODIS_AIRS.ppt application/vnd.ms-powerpoint R24KLSE7EC7LFVAIFK6MOV2P4E7EHIBL Slide 1 - 1601-01-01T00:00:00Z 2015-03-16T21:22:21Z 172 13 472 1455104 3f3617ddfde34eba50e3877b4fdc2649625a2767841a65314797a4fa7f01897c 7f5d35fb9483a0dc519167a9d076254a9f76f36f443287262c4dbf3b6e64087fea910e1bea3005

In [13]:
print(linecount)

1001


In the cell above, we receive 1,001 as the line count since the variable indicates a count for all the lines, including the headers. So, this is what we would expect if the set includes 1,000 entries.

Now, let's read the dataset into a dictionary for easier processing.

In [19]:
lcwa_powerpoint_info = list()

with open('lcwa_gov_powerpoint_metadata.csv', 'r', newline='', encoding='utf-16') as data:
    lcwa_powerpoint_data = csv.DictReader(data, dialect='excel-tab', delimiter='\t')
    for line in lcwa_powerpoint_data:
        lcwa_powerpoint_info.append(line)

In [21]:
lcwa_powerpoint_info[:2]

[OrderedDict([('urlkey', 'ada.ky.gov/adaemploymenttaxincentives.ppt'),
              ('timestamp', '20100331231723'),
              ('original', 'http://ada.ky.gov/ADAEMPLOYMENTTaxIncentives.ppt'),
              ('mimetype', 'application/vnd.ms-powerpoint'),
              ('digest', '4Y6Z6KRRHS3TIVQ2ZJVYVW7RSGXOBS6Y'),
              ('title',
               '   Human Development Institute Promoting Independence, Productivity, and Integration for All People _________________________________   '),
              ('company', '-'),
              ('creation_date', '1998-12-17T14:31:23Z'),
              ('last_modified', '2005-03-25T15:56:39Z'),
              ('revision_number', '47'),
              ('slide_count', '31'),
              ('word_count', '781'),
              ('file_size', '137216'),
              ('sha256',
               '9b6b74447622ac1646a538e63e452569e111cf775d1244373678d947405e33be'),
              ('sha512',
               '9567992755c0b1b38fdccb4440a4bb4b1758301b5cd1978b9

In [37]:
pptx_headers = list()
for element in lcwa_powerpoint_info[0]:
    pptx_headers.append(element)

In [38]:
pptx_headers

['urlkey',
 'timestamp',
 'original',
 'mimetype',
 'digest',
 'title',
 'company',
 'creation_date',
 'last_modified',
 'revision_number',
 'slide_count',
 'word_count',
 'file_size',
 'sha256',
 'sha512']

In [43]:
creators = dict()

for ppt in lcwa_powerpoint_info:
    if ppt['company'] in creators:
    

-
Lister Staveley-Smith
State of Delaware
SC Office of Human Resources
Fermilab PPD
CERN
HRSA
CDC
NCHSTP
UC Davis
ICPSR

Southface Energy Institute

Accenture
State of Washington, Department of Personnel
-
GSA
Virginia Beach Public Schools
NIH
DOT
Fire Safety Section, FAA
US Treasury - FMS
Low + Associates
-
GSA
GSA
Center for Health Statistics
USDA - NRCS
DoED
U.S. Department of Education
Fermilab
VITA
-
-
DOER
Bull Services
St. John Evangelist
Department of Communications
Home
LDCT
-
-
USDA
Tippecanoe County
<DIT>
USDA
SLAC
NASA/GSFC
LAP-AUTH
-
California Army National Guard
Lockheed Martin Information Technology
-
CDC
Department of Consumer Affair
State of California
-
GSFC/NASA
Hooper Graphic Design/Perfect Mix
Dell Computer Corporation
Department of Personnel
sde
Hewlett-Packard
CA Energy Commission
US FDA
CoosWatershed
-
DOH
LMIT-ODIN
University of Georgia
USGS/EROS Data Center
USRA
	閠]狴逄嬘뿿��
UNC-CH
Library of Congress
loc
-
NAPHSIS
Terberg Design
-
USDA/ARS NSTL
Emory University

##################
Using pandas

In [24]:
ppt_df = pd.DataFrame.from_records(lcwa_powerpoint_info)

In [25]:
ppt_df.head()

urlkey       timestamp  \
0          ada.ky.gov/adaemploymenttaxincentives.ppt  20100331231723   
1  astrophysics.gsfc.nasa.gov/conferences/superno...  20090419154246   
2  awm.delaware.gov/sitecollectiondocuments/awm+g...  20100220062758   
3  bcbintranet.sc.gov/ohr/hr-advisory/workforcepl...  20090416103442   
4  beamdocs.fnal.gov/ad/docdb/0010/001026/003/040...  20151115084934   

                                            original  \
0   http://ada.ky.gov/ADAEMPLOYMENTTaxIncentives.ppt   
1  http://astrophysics.gsfc.nasa.gov/conferences/...   
2  http://www.awm.delaware.gov/SiteCollectionDocu...   
3  http://www.bcbintranet.sc.gov/OHR/hr-advisory/...   
4  http://beamdocs.fnal.gov/AD/DocDB/0010/001026/...   

                        mimetype                            digest  \
0  application/vnd.ms-powerpoint  4Y6Z6KRRHS3TIVQ2ZJVYVW7RSGXOBS6Y   
1  application/vnd.ms-powerpoint  7EZSHVTKZE2FXP5SXI6DO2RJPIDLMXOP   
2  application/vnd.ms-powerpoint  ZHT4MJOX7D7OIRMCBPX4PMWZ4OQ7XSZ2   
3  application/vnd.ms-powerpoint  KDBPF6PJCBK2BEIYC2L5I33THZMINFML   
4  application/vnd.ms-powerpoint  OJOHGFEE4EI44UNSXLNV4Q5IKJ2L4H2Y   

                                               title  \
0     Human Development Institute Promoting Indep...   
1                        xNTD and the Pathway to SKA   
2                                   VSM Data Metrics   
3  Why Workforce Planning is Important during Har...   
4                            PowerPoint Presentation   

                        company         creation_date         last_modified  \
0                             -  1998-12-17T14:31:23Z  2005-03-25T15:56:39Z   
1         Lister Staveley-Smith                     -  2007-02-20T13:41:45Z   
2             State of Delaware  2006-10-16T16:19:52Z  2006-10-26T19:37:20Z   
3  SC Office of Human Resources  2009-01-07T14:28:04Z  2009-01-09T16:31:58Z   
4                  Fermilab PPD  2003-06-30T15:15:33Z  2004-02-22T22:33:55Z   

  revision_number slide_count word_count file_size  \
0              47          31        781    137216   
1              51          17        553   1580032   
2               4           1          2    136704   
3               7           9        238     53760   
4             122          45       1638   8227840   

                                              sha256  \
0  9b6b74447622ac1646a538e63e452569e111cf775d1244...   
1  34c77d3dc3cd3dc51d084584cafedf24189a42bf6bf6d3...   
2  1db9d640446ab52654f891458a30c6219e0125915f61c4...   
3  758efaac57fdf38d82bf7271f9cd5d1f083f842d884bff...   
4  3aeff996f66ee7f51e045ede1f5d21ab57639928b35804...   

                                              sha512  
0  9567992755c0b1b38fdccb4440a4bb4b1758301b5cd197...  
1  c36c59e694aad68c7b5fd4b32383e857520778131d9010...  
2  caa579649ea4a0a3126ab9842fe11e6e548695132d0da1...  
3  19449593a8c1defae33567966314bfe72d822b387eac6b...  
4  c6124015a81adf4d41e563b0a6460fd20275744f12b9b8...

In [45]:
creators = ppt_df.groupby('company')

creators.head()

urlkey       timestamp  \
0            ada.ky.gov/adaemploymenttaxincentives.ppt  20100331231723   
1    astrophysics.gsfc.nasa.gov/conferences/superno...  20090419154246   
2    awm.delaware.gov/sitecollectiondocuments/awm+g...  20100220062758   
3    bcbintranet.sc.gov/ohr/hr-advisory/workforcepl...  20090416103442   
4    beamdocs.fnal.gov/ad/docdb/0010/001026/003/040...  20151115084934   
5    beamdocs.fnal.gov/ad/docdb/0022/002206/001/shi...  20151114112127   
6                   bhpr.hrsa.gov/grants00/ta-2000.pps  20010429074551   
7        cdc.gov/nchstp/dstd/stats_trends/chlamydi.ppt  20060828091122   
8    cdc.gov/tb/programs/evaluation/guide/webinar/u...  20091113094734   
9    cdph.ca.gov/programs/wicworks/documents/babybe...  20110703051408   
10   digitalpreservation.gov/meetings/documents/ndi...  20140108050503   
11   digitalpreservation.gov/news/events/ndiipp_mee...  20090902133053   
12   dnr.sc.gov/marine/nerr/present/sustainable/jef...  20090817072049   
13   doleta.gov/performance/trainingtutorials/ppt/w...  20120804204306   
14   dop.wa.gov/sitecollectiondocuments/payroll/hrm...  20101027101013   
15   dop.wa.gov/sitecollectiondocuments/payroll/hrm...  20090922150030   
16                    dot.ri.gov/contracting/quest.ppt  20090416044315   
17   ebuy.gsa.gov/images/advrewrite/tutorial/orderh...  20100819001809   
18   ercm.ed.gov/views/documents/communication_pont...  20091212195417   
19   ethics.od.nih.gov/topics/non-fte-trainee-activ...  20090610203741   
20   fhwa.dot.gov/planning/freight_planning/talking...  20170321023542   
21    fire.tc.faa.gov/ppt/systems/lithsysfire12-00.ppt  20011121105253   
22   fms.treas.gov/gtas/agency_outreach_presentatio...  20090424033138   
23   fsa4counselors.ed.gov/clcf/attachments/finding...  20090320021718   
24   goes-r.gov/downloads/goes+users'+conference+ii...  20151111232929   
25   gsa.gov/graphics/admin/ae_acquisition_process_...  20101128073912   
26   gsa.gov/graphics/pbs/abaas_training_slides_jan...  20101101055751   
27   healthy.arkansas.gov/programsservices/healthst...  20101128053636   
28   ias.sc.egov.usda.gov/help/portal/06_data_sourc...  20100708190424   
29   ies.ed.gov/director/conferences/06ies_conferen...  20091222223658   
..                                                 ...             ...   
947  indico.fnal.gov/getfile.py/access?contribid=5&...  20151203062017   
948  os.hhs.gov/healthit/ahic/materials/meeting11/c...  20090417221054   
949  beamdocs.fnal.gov/ad/docdb/0014/001490/002/top...  20151117192055   
950  cdph.ca.gov/certlic/occupations/documents/rehs...  20110218134930   
953  ncbi.nlm.nih.gov/bookshelf/picrender.fcgi?book...  20100615090501   
954  cdph.ca.gov/healthinfo/healthyliving/childfami...  20110308123039   
955  cdph.ca.gov/programs/mcah/documents/mo-overvie...  20090501144516   
956  consensus.nih.gov/2009/fhx+images/conference+s...  20090618042659   
957  indico.fnal.gov/getfile.py/access?contribid=8&...  20151115015502   
962  ed.gov/about/offices/list/ovae/pi/hs/04summit/...  20170323112945   
963  elandings.alaska.gov/confluence/download/attac...  20100308042306   
964  ffas.usda.gov/icd/stconf/event5/ggordillo%5B1%...  20100309020130   
965  hss.doe.gov/deprep/facrep/workshop2004/present...  20100203095816   
967  osha.gov/dep/fap/fshc58/2003-council_awards_pr...  20170707174312   
970  cdph.ca.gov/programs/cpns/documents/network-fv...  20110823083838   
972                     hcfa.gov/medicare/fulldata.ppt  20000930224750   
974  p25ext.lanl.gov/people/mbrooks/silicon/unmmeet...  20151115061658   
976  cdph.ca.gov/programs/cpns/documents/nsc-health...  20111223042830   
977  cdph.ca.gov/programs/immunize/documents/hcw+im...  20110213232428   
984  cdph.ca.gov/certlic/occupations/documents/rehs...  20120316031049   
987  csm.ornl.gov/workshops/biophysics10/documents/...  20151201150536   
988  fgdc.gov/ngac/meetings/august-2009/public-priv...  20110515143912   
989                  hep.anl.gov/repond/xia_092807.p